### PR TITLE
modules/keymap: include buffer when registering `keymapsOnEvents`

### PR DESCRIPTION
**Pull request by jfly**
This feels more correct to me: I believe we were registering a keymap for all buffers, even if only some of them have experienced the relevant autocmd.

I don't know if it's guaranteed here that `args.buf` always exists (but maybe it's not a problem if it doesn't exist, because we'll still register a keymap for all buffers?).

This fixes https://github.com/nix-community/nixvim/issues/2711

### CODE REVIEW

**Review by MattSturgeon — APPROVED** [review]
Makes sense to me, but I'll hold off for a second opinion @nix-community/nixvim

**Review by traxys — APPROVED** [review]

**Comment by MattSturgeon** [comment]
@mergify queue

**Comment by jfly** [comment]
Thanks for the quick merge, @matt

**Comment by khaneliman** [comment]
Bisecting a lua error points this commit as the cause. 

```bash
Error executing lua callback: vim/shared.lua:0: key found in more than one map: buffer
stack traceback:
	[C]: in function 'error'
	vim/shared.lua: in function 'tbl_extend'
	/nix/store/d7h1rgpfh4ds52p7mj71gpj1yvgs00cg-init.lua: in function </nix/store/d7h1rgpfh4ds52p7mj71gpj1yvgs00cg-init.lua:0>
	[C]: in function 'nvim_exec_autocmds'
	...ompiled-0.10.2/share/nvim/runtime/lua/vim/lsp/client.lua: in function '_on_attach'
	...ompiled-0.10.2/share/nvim/runtime/lua/vim/lsp/client.lua: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
```

**Comment by jfly** [comment]
@khaneliman, I opted to use `vim.tbl_extend("error", ...)` to try to avoid suprising people that we're now specifying a buffer for them.

Are you specifying an `autoCmd` that specifies a buffer?

**Comment by khaneliman** [comment]
I've got only a handful of autoCmds https://github.com/search?q=repo%3Akhaneliman%2Fkhanelivim%20autoCmd&type=code and then a bunch of plugins with will create their own.

**Comment by jfly** [comment]
I guess the easiest way to track this down would be to look at your generated init.lua file. Can you give me a command to repro the issue?

**Comment by MattSturgeon** [comment]
Hm, maybe extending with "error" mode isn't the best approach? If a user has explicitly configured something, that should override whatever defaults we're setting. :thinking:

**Comment by jfly** [comment]
> If a user has explicitly configured something, that should override whatever defaults we're setting.

Generally I'd agree, but in this case, I think we're talking about a buffer hardcoded at build time? Which doesn't make any sense to me. Perhaps I'm fundamentally misunderstanding what's happening here. Having a repro would be nice to set me straight.

**Comment by MattSturgeon** [comment]
> a buffer hardcoded at build time

I believe you can use "buffer 0" to reference the "current" buffer? So maybe some people are explicitly doing that?

I haven't looked too closely, so it's likely I'm missing something fundamental too. But my first thought is that if we don't want to allow setting buffnr through nixvim options, that should be a eval time filter and/or warning/assertion.

**Comment by jfly** [comment]
Ah yeah, using "buffer 0" could make some sense.

Sorry for the breakage. I'm leaving for a trip soon, and won't have any real time to spend on this until the new year. Here's a quickfix: https://github.com/nix-community/nixvim/pull/2719.

**Pull request merged**
